compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+from xdsl.context import MLContext
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.dialects.builtin import IndexType
+from xdsl.ir import Block, Operation, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+from xdsl.traits import Pure
+
+
+def find_parent_for_loop(op: Operation) -> scf.For | None:
+    """
+    Find the parent for-loop of the operation.
+    If no for-loop is found, return None.
+    """
+    op = op.parent_op()
+    if op is None:
+        return None
+    while not isinstance(op, scf.For):
+        if op.parent_op() is None:
+            return None
+        op = op.parent_op()
+    return op
+
+
+def is_in_loop(op: Operation) -> bool:
+    """
+    Check if the operation is inside a loop.
+    """
+    return find_parent_for_loop(op) is not None
+
+
+def defined_outside_loop(op: Operation) -> bool:
+    """
+    Check if all operands of the operation are defined outside the loop.
+    """
+    for operand in op.operands:
+        if find_parent_for_loop(operand.owner) is find_parent_for_loop(op):
+            return False
+    return True
+
+
+class LowerPureOperations(RewritePattern):
+    """
+    This class represents a rewrite pattern for moving any Operation outside
+    a (nested) for-loop. An Oparation can be moved if it is pure, that is, it
+    does not have any side-effects within the loop and is not dependent on any loop variable.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, main_op: Operation, rewriter: PatternRewriter):
+        def can_move_operation(op: Operation) -> bool:
+            """
+            Operations can be moved if they comply with the following conditions:
+            1. The operation is inside a loop.
+            2. All operands are defined outside the loop.
+            3. The operation is Pure, it doesnt have any side effects.
+            """
+            if all(
+                [
+                    is_in_loop(op),
+                    defined_outside_loop(op),
+                    Pure() in op.traits or isinstance(op, memref.Alloc),
```

Comment:
I decided not to split the rewrite patterns for the alloc case and the general case as all almost code is equal. The only difference lies here in the testing if it can be moved, so I just worked with an or-operation to cover both cases. If we were to push the general case upstream to xDSL, the alloc case needs te be removed

Reply:
Ok! Makes sense

Reply:
@JonasCrols, maybe instead, make the memref.Alloc a parameter of this rewrite pattern? Then you can make it more general? I believe I've seen @AntonLydike do this a couple of times.